andela/ah-backend-prime
# #164798287 Implement JWT Authentication

#### What does this PR do?
- Send a JSON Web Token(JWT) to a user upon successful registration in order to access other API endpoints that require authentication.

#### Description of tasks to be completed
- Add JWTAuthentication class to authenticate users
- Add a method that generates a token
- Add token as field in serializers

#### How should this be manually tested?
- Fetch this branch by running the command `git fetch origin ft-users-receive-jwt-164798287` in the terminal.
- Check into this branch by running `git checkout ft-users-receive-jwt-164798287`
- Activate your virtual environment by running `source venv/bin/activate`
- Delete the current database by running drop database DB_NAME; in the psql shell
- Create a new database by running create database DB_NAME; in the psql shell
- Update the .env file with the database name
- Source the .env file with the updated database name
- Migrate by running python manage.py migrate
- Start the application by running python manage.py runserver
- To test the register user API endpoint, Open postman, choose the POST method and paste this registration url http://localhost:8000/api/users/register/
- Enter raw JSON data in the format below and send the request.
```
{ "user" : { 
      "email": "emmanuel@primemail.com", 
      "username": "holiness",  
      "password": "holinesssupreme"
  }
}
```
- After successful registration, a response should be sent in the format below with a token included.
```
{
    "user": {
        "email": "emmanuel@primemail.com",
        "username": "holiness",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MiwiZW1haWwiOiJlbW1hbnVlbEBwcmltZW1haWwuY29tIiwiZXhwIjoxNTU0MTQ5NzI4fQ.OL5eihYRBZ1OB_CnvZotoD4N9-a_WH3hrj1Inylo_gw"
    }
}
```
- To test the Login API endpoint, Open postman, choose the POST method and paste this login url http://localhost:8000/api/users/login/.
- Enter raw JSON data in the format below and send the request
```
{ "user" :  { 
      "email": "emmanuel@primemail.com", 
      "password": "holinesssupreme" 
  }
}
```
- After successful login, a response should be sent in the format below with a token included.
```
{
    "user": {
        "email": "emmanuel@primemail.com",
        "username": "holiness",
        "token": "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpZCI6MiwiZW1haWwiOiJlbW1hbnVlbEBwcmltZW1haWwuY29tIiwiZXhwIjoxNTU0MTQ5OTAyfQ.LULCTPHOwE_xqrJblzsVCD8iAu4Y9d2B3CwmtZFTjUQ"
    }
}
```
#### What are the relevant pivotal tracker stories?
[#164798287](https://www.pivotaltracker.com/story/show/164798287)

#### Screenshots( where appropriate)
- Registering a user
![image](https://user-images.githubusercontent.com/36891299/55349699-0d4cbe80-54c3-11e9-9e45-687dd3b6ad64.png)

- Logging in a registered user
![image](https://user-images.githubusercontent.com/36891299/55349851-5ef54900-54c3-11e9-81ba-a9bf15665436.png)




[Finishes #164798287]
